### PR TITLE
[VPU] Improving proposal testing methods

### PR DESCRIPTION
**Pull request by akruglov18**
### Details:
Improving proposal testing methods for MyriadX plugin.
 The method creates a bipartite graph according to the following rules:
1)The vertices of first part are rectangles of optimized version,
the vertices of second part are rectangles of reference version.
2)There is an edge between two vertices if the rectangles corresponding
to these vertices are equal. The criterion for equality
is intersection over union.
The task is reduced to finding the maximum matching in the graph.
The Kuhn's algorithm is used to find the maximum matching.
The method using the maximum matching in the graph worked correctly in most test cases than current method. This is why this method should be used.
Calculation IoU wasn't correct for normalized data. Added special function for calculating.
Correctness checking of coordinates. There is more information about incorrect data.
Added an optional scores checking without threshold.
Added an optional soft scores checking. Scores checking element-by-element with threshold.

### CODE REVIEW

**Comment by akruglov18** [comment]
@gladilov-gleb, please review

**Comment by gladilov-gleb** [comment]
@akruglov18 is it ready for review?

**Comment by akruglov18** [comment]
> @akruglov18 is it ready for review?

Yes.

File: inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+        //edges from vertexes of optimized boxes to vertexes of reference boxes
+        Graph graph {make_graph(reference, count_reference, optimized, count_optimized,
+                                reference_scores, output_scores,
+                                normalized, withOutputScores)};
+
+        int matching_count = 0;
+        std::vector<bool> used(count_optimized, false);
+        std::vector<int> mt(count_reference, -1); // matching, -1 means there is no matching
+        for (std::size_t vertex = 0; vertex < count_optimized; ++vertex) {
+            if (try_kuhn(used, mt, graph, vertex)) {
```
Comment (gladilov-gleb) [inline inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp:100]: Interface looks a bit strange: `try_kuhn` takes `mt` as an argument, but this is not changed anywhere outside of function - looks like `mt` is some internal container for algorithm. If so it is better to encapsulate it inside class.

File: inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp [hunk 2]
```diff
@@ -0,0 +1,427 @@
...
+    }
+
+private:
+    float _threshold;
+
+    Graph make_graph(const float* reference, std::size_t count_reference,
+                     const float* optimized, std::size_t count_optimized,
+                     const float* reference_scores, const float* output_scores,
+                     bool normalized, bool withOutputScores) {
+        Graph g(count_optimized);
```
Comment (gladilov-gleb) [inline inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp:119]: Don't use one-letter names

File: inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp [hunk 3]
```diff
@@ -0,0 +1,427 @@
...
+        validateData(reference, count_reference, optimized, count_optimized);
+
+        //edges from vertexes of optimized boxes to vertexes of reference boxes
+        Graph graph {make_graph(reference, count_reference, optimized, count_optimized,
+                                reference_scores, output_scores,
+                                normalized, withOutputScores)};
+
+        int matching_count = 0;
+        std::vector<bool> used(count_optimized, false);
+        std::vector<int> mt(count_reference, -1); // matching, -1 means there is no matching
```
[the remark this answers is not in the document]
Comment (akruglov18) [reply inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp:98]: mt is used to save matching. For example mt[2] = 1. This means that we came to vertex number 2 of the second part from vertex number 1 of the first part.

File: inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp [hunk 4]
```diff
@@ -0,0 +1,427 @@
...
+
+using namespace InferenceEngine;
+using namespace PrecisionUtils;
+
+/**
+* "IoU = intersection area / union area" of two boxes A, B
+*   A, B: 4-dim array (x1, y1, x2, y2)
+*/
+
+static float check_iou(const float* A, const float* B) {
```
[the remark this answers is not in the document]
Comment (akruglov18) [reply inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp]: fixed

File: inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp [hunk 5]
```diff
@@ -0,0 +1,427 @@
...
+    if (A[0] > B[2] || A[1] > B[3] || A[2] < B[0] || A[3] < B[1]) {
+        return 0.0f;
+    } else {
+        // overlapped region (= box)
+        const float x1 = std::max(A[0], B[0]);
```
[the remark this answers is not in the document]
Comment (akruglov18) [reply inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp]: fixed

File: inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp (hunk 1 above)
Comment (akruglov18) [reply inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp]: fixed

File: inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp (hunk 2 above)
Comment (akruglov18) [reply inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp]: Fixed

File: inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp [hunk 6]
```diff
@@ -0,0 +1,427 @@
...
+        result << "ratio = " << CommonTestUtils::vec2str(ratio) << "_";
+        result << "scale = " << CommonTestUtils::vec2str(scale) << "_";
+        result << "clip_before_nms=" << clip_before_nms << "_";
+        result << "clip_after_nms=" << clip_after_nms << "_";
+        result << "normalize=" << normalize << "_";
+        result << "box_size_scale=" << box_size_scale << "_";
+        result << "box_coordinate_scale=" << box_coordinate_scale << "_";
+        result << "framework=" << framework << "_";
+
+    return result.str();
```
[the remark this answers is not in the document]
Comment (akruglov18) [reply inference-engine/tests/functional/plugin/myriad/single_layer_tests/proposal.cpp]: Fixed

**Review by gladilov-gleb — APPROVED** [review]

**Pull request closed without merging**